third_party/move/move-compiler-v2/src/pipeline/dead_store_elimination.rs

```diff
@@ -1,57 +1,202 @@
+    dead: BTreeSet<CodeOffset>,
+}
+
+impl DefUseGraph {
+    /// Create a new `DefUseGraph` from the function `target`.
+    pub fn new(target: &FunctionTarget) -> Self {
+        let mut this = Self {
+            children: BTreeMap::new(),
+            parents: BTreeMap::new(),
+            dead: BTreeSet::new(),
+        };
+        this.populate_from(target);
+        this
+    }
+
+    /// Obtain the set of dead stores, i.e., code offsets which can be removed safely.
+    pub fn dead_stores(mut self) -> BTreeSet<CodeOffset> {
+        let mut dead = BTreeSet::new();
+        while let Some(offset) = self.remove_a_dead_node() {
+            dead.insert(offset);
+        }
+        dead
+    }
+
+    /// Populate an empty graph from the (restricted) definitions and uses in `target`.
+    fn populate_from(&mut self, target: &FunctionTarget) {
+        let code = target.get_bytecode();
+        let live_vars = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live variable annotation is a prerequisite");
+        for (offset, instr) in code.iter().enumerate() {
+            use Bytecode::*;
+            match instr {
+                Assign(_, dst, src, _) if dst == src => {
+                    // self-assignment is always a dead store
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, true);
+                },
+                Assign(_, dst, ..) | Load(_, dst, _) => {
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, false);
+                },
+                _ => {},
+            }
+        }
+    }
+
+    /// Remove one dead node (arbitrary but deterministic) from the graph, if present.
+    fn remove_a_dead_node(&mut self) -> Option<CodeOffset> {
+        if let Some(node) = self.dead.pop_last() {
+            let parents = self.disconnect_from_parents(node);
+            let children = self.disconnect_from_children(node);
+            // Reconnect all parents to all children.
+            for parent in parents.iter() {
+                for child in children.iter() {
+                    self.children.entry(*parent).or_default().insert(*child);
+                    self.parents.entry(*child).or_default().insert(*parent);
+                }
+            }
+            // Parents are the only ones who could become dead (if `node` was their last child).
+            parents
+                .iter()
+                .for_each(|parent| self.re_evaluate_death(*parent));
+            Some(node)
+        } else {
+            None
+        }
+    }
+
+    /// Check if a `parent` node is now dead, and mark it as such if it is.
+    /// A parent node is always a (restricted) definition, and is dead if all its children are dead.
+    fn re_evaluate_death(&mut self, parent: CodeOffset) {
+        if let Some(children) = self.children.get(&parent) {
+            if children.iter().all(|child| self.dead.contains(child)) {
```

Comment:
Should dead children be still connected to their parents?

Reply:
It is safe to do so, as they will eventually be removed (the algo was written to remove one dead node at a time). I have re-written the algo to remove dead stuff bottom up.